### ISSUE TITLE
paste does not work in firefox

### ISSUE THREAD

**Issue by AdamRollinson** [danialfarid/ng-file-upload#1183]
CTRL + V is not working in firefox (42.0)

any work around for this?

works fine in chrome.

**Comment by danialfarid** [comment]
Since version 10.1.9 add `ngf-enable-firefox-paste="true"` to enable paste in firefox.
The trick is to make the input as `contenteditable` but disable editing the content and just capturing the paste events.

**Comment by AdamRollinson** [comment]
Still does not work.

```
<div ng-click="notify()" style="margin-top: 10px;" ngf-drop="" ng-model="file" class="drop-box ng-pristine ng-untouched ng-valid ng-empty" ngf-drag-over-class="'dragover'" ngf-capture="capture" accept="image/*" ngf-pattern="'image/*'" ngf-enable-firefox-paste="true">Upload</div>
```

Am I doing something wrong?

**Comment by danialfarid** [comment]
If you copy your code into demo page it works on Firefox for me.

**Comment by AdamRollinson** [comment]
It works when you copy an image from the web but if you are copying from the snipping tool for example, does not work.

**Comment by pkaminski** [comment]
Pasting images in Firefox is broken for me as well.  I'm installing `ngf-drop` on the parent element of a `textarea`, which also kills pasting text (due to the `preventDefault` on `paste` events that `ng-file-upload` implements in Firefox).  When I add `ngf-enable-firefox-paste="true"`, pasting images still doesn't work in Firefox because the focus remains in the `textarea`, but on top of that all keyboard events are eaten so I can't even type into the field!

I've worked around this by patching `preventDefault` on `paste` events, disabling paste handling in `ngf-file-upload`, and using https://github.com/layerssss/paste.js instead, which works pretty well (though it requires some extra contortions to make text pasting work in Firefox, and undo/redo of pastes gets wonky).  Perhaps you can borrow some ideas from that project and integrate them into `ng-file-upload`?

**Comment by danialfarid** [comment]
@AdamRollinson 
Pasting from file system does not work in any browser. The paste feature is for in browser paste, so you can copy the image from any tab in the browser and paste it in the area, but not from file system.

**Comment by danialfarid** [comment]
@pkaminski Could you create a jsfiddle and explain what is not working?
You could have the image paste on a separate div so it doesn't interfere with the textarea.

**Comment by AdamRollinson** [comment]
@danialfarid
it works on chrome from filesystem.

**Comment by danialfarid** [comment]
@AdamRollinson what operating system?

**Comment by AdamRollinson** [comment]
@danialfarid 
Windows 10

**Comment by pkaminski** [comment]
@danialfarid Check out http://plnkr.co/edit/i41Yx5MtRFRZFVVmMaSz?p=preview.  It works fine in Chrome, but on Firefox pasting of both text and images fails.  If you add `ngf-enable-firefox-paste="true"` then typing stops working as well.

Perhaps I misunderstand something, but doesn't the `ngf-drop` need to be on an ancestor of the `textarea`?  Otherwise, how would it receive the `paste` event?  Maybe you could include an example of how to use `ngf-drop` on a `textarea` in the project's `README`.

**Comment by danialfarid** [comment]
@pkaminski the example doesn't ready work in Chrome:
http://plnkr.co/edit/wUt6wVkqfHSbl9xOOAfY?p=preview
It doesn't assign the pasted file to the model.

**Comment by pkaminski** [comment]
@danialfarid My original example does work in Chrome.  I don't care about a model -- I just use the callback directly, passing in `$file`.  Also, your version is actually broken in Firefox in the same way: you can't paste either text or images in the textarea.

**Comment by danialfarid** [comment]
@pkaminski For now I would suggest use separate divs for text area and image paste. Text area is for text and doesn't make much sense to paste image in it. So you could have the image paste div separate from textarea and it should work in both.

**Comment by pkaminski** [comment]
> Text area is for text and doesn't make much sense to paste image in it.

@danialfarid On the contrary, I think that's a very common context for pasting images.  You have to look no further than the GitHub comment area you're using right now for an example...  Anyway, if `ng-file-upload` isn't meant to support file uploading through a text area, you might at least want to note it prominently in the `README`.  But as I mentioned, it's really not far off, if you could just adopt some code from https://github.com/layerssss/paste.js.

**Comment by danialfarid** [comment]
Correct, but pasting images on github is only working in Chrome not Firefox. 
I suggested that approach as temporary solution. I will have a look to see if it is fixable later on but it is kind of a low priority considering it is just a hack for one browser.
If you are willing to contribute PRs are welcome.

**Comment by pkaminski** [comment]
> Correct, but pasting images on github is only working in Chrome not Firefox. 

True, but 1) it can be made to work in Firefox too (see my product, https://reviewable.io), though there are some trade-offs, and 2) even if you decide to only support it in Chrome it should at least not break Firefox textareas altogether.

I don't really have the time to dig into `ng-file-upload` code to make a clean patch (sorry!) but the key insight from `paste.js` in non-Chrome browsers is simply to intercept `{ctrl,cmd}-v` and switch the focus to an off-screen (and out-of-hierarchy) `contenteditable` to get a `paste` event, then switch focus back to the text box afterwards.  Then you don't need any hacks to make the `contenteditable` not actually editable and everything is happy.

**Comment by danialfarid** [comment]
Fix firefox paste text in textarea inside ngf-drop.

**Comment by danialfarid** [comment]
firefox paste text in textarea inside ngf-drop is fixed at version 10.1.11

**Comment by danialfarid** [comment]
`ngf-enable-firefox-paste` would be experimental for now until I find a way to integrate it better probably by using paste.js.

**Comment by xavier-figueroa** [comment]
I just want to let you know that I'm testing the latest version of your plugin and it is not working in FF

**Comment by danialfarid** [comment]
@xavier-figueroa what is not working?

**Comment by xavier-figueroa** [comment]
in https://angular-file-upload.appspot.com/ the paste functionality only works in chrome. I integrated it in my application with the same results. I installed it using bower

**Comment by danialfarid** [comment]
yes that's what this issue is about. You can enable it by https://github.com/danialfarid/ng-file-upload/issues/1183#issuecomment-177043394 but experimental
